src/gil.rs

```diff
@@ -276,6 +282,17 @@ impl ReferencePool {
     }
 }
 
+#[inline]
+pub(crate) fn update_deferred_reference_counts(py: Python<'_>) {
+    // Justification for relaxed: worst case this causes already deferred drops to be
+    // delayed slightly later, and this is also a one-time flag, so if the program is
+    // using deferred drops it is highly likely that branch prediction will always
+    // assume this is true and we don't need the atomic overhead.
+    if POOL.ever_used.load(sync::atomic::Ordering::Relaxed) {
+        POOL.update_counts(py)
+    }
+}
+
```

Comment:
If we do this, I would prefer to better encapsulate it, e.g. in `impl ReferencePool`

```rust
#[inline]
fn update_counts(&self, _py: Python<'_>) {
   if self.ever_used.load(Ordering::Relaxed) {
     self.update_counts_impl(py);
   }
}

#[inline(never)]
fn update_counts_impl(&self, _py: Python<'_>) { .. }
```

Reply:
We might even get away with not implementing all of this ourselves at all and get better self-documenting code: Just put it inside a `once_cell::sync::Lazy`, i.e.

```rust
static POOL: Lazy<ReferencePool> = Lazy::new(ReferencePool::new);
```

and call `register_decref` as before (going via `impl Deref for Lazy`, i.e. `Lazy::force`), but for update, do

```rust
if let Some(pool) = POOL.get() {
  pool.update_counts();
}
```

Reply:
(This would also take care of not updating the flag after the first change, but it might be a bit slower because `Lazy` track initialization state inside a `AtomicU8`.)

Reply:
Yes it's a good idea, I'll test out performance of using `Lazy` when I get a chance later 👍